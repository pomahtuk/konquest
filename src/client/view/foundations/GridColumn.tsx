import React, { ReactElement, useContext, HTMLAttributes, ReactNode } from "react";
import { styled } from "styletron-react";
import { ThemeContext } from "../themes/ThemeProvider";

export type GridSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | "half" | "full";

export type GridColumnProps = {
  children: ReactNode;
  size?: GridSize;
  sizeMedium?: GridSize;
  sizeLarge?: GridSize;
  sizeHuge?: GridSize;
  offset?: GridSize;
  offsetMedium?: GridSize;
  offsetLarge?: GridSize;
  offsetHuge?: GridSize;
  className?: string;
  attributes?: HTMLAttributes<HTMLDivElement>;
  bleed?: boolean;
};

const toRealValue = (val?: GridSize): GridSize | undefined => (val === "half" ? 6 : val === "full" ? 12 : val);

const GridColumn = ({
  size,
  sizeMedium,
  sizeLarge,
  sizeHuge,
  children,
  className,
  attributes,
  offset,
  offsetMedium,
  offsetLarge,
  offsetHuge,
  bleed
}: GridColumnProps): ReactElement => {
  const theme = useContext(ThemeContext);
  // prepare variables
  const realSize = toRealValue(size) || 12;
  const realOffset = toRealValue(offset);
  const realSizeMedium = toRealValue(sizeMedium);
  const realSizeLarge = toRealValue(sizeLarge);
  const realSizeHuge = toRealValue(sizeHuge);
  const realOffsetMedium = toRealValue(offsetMedium);
  const realOffsetLarge = toRealValue(offsetLarge);
  const realOffsetHuge = toRealValue(offsetHuge);
  // prepare styled component
  const GridColumnStyled = styled("div", {
    boxSizing: "border-box",
    padding: bleed ? "0" : `${theme.units.medium} calc(${theme.units.medium}/2) 0 calc(${theme.units.medium}/2)`,
    flexBasis: `calc(100% / (12/${realSize}))`,
    width: `calc(100% / (12/${realSize}))`,
    maxWidth: `calc(100% / (12/${realSize}))`,
    ...(realOffset ? { marginLeft: `calc(100% / (12/${realOffset}))` } : {}),
    [theme.media.medium]: {
      padding: bleed ? "0" : `${theme.units.large} calc(${theme.units.large}/2) 0 calc(${theme.units.large}/2)`,
      ...(realSizeMedium ? { width: `calc(100% / (12/${realSizeMedium}))` } : {}),
      ...(realOffsetMedium ? { marginLeft: `calc(100% / (12/${realOffsetMedium}))` } : {})
    },
    [theme.media.large]: {
      padding: bleed ? "0" : `${theme.units.larger} calc(${theme.units.larger}/2) 0 calc(${theme.units.larger}/2)`,
      ...(realSizeLarge ? { width: `calc(100% / (12/${realSizeLarge}))` } : {}),
      ...(realOffsetLarge ? { marginLeft: `calc(100% / (12/${realOffsetLarge}))` } : {})
    },
    [theme.media.huge]: {
      padding: bleed ? "0" : `${theme.units.largest} calc(${theme.units.largest}/2) 0 calc(${theme.units.largest}/2)`,
      ...(realSizeHuge ? { width: `calc(100% / (12/${realSizeHuge}))` } : {}),
      ...(realOffsetHuge ? { marginLeft: `calc(100% / (12/${realOffsetHuge}))` } : {})
    }
  });
  return (
    <GridColumnStyled {...attributes} className={className}>
      {children}
    </GridColumnStyled>
  );
};

export default GridColumn;
