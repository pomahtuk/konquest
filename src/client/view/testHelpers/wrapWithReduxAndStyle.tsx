import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
