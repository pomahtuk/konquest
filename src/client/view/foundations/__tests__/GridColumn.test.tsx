import React from "react";
import { render } from "@testing-library/react";
import { matchers } from "jest-emotion";

expect.extend(matchers);

import wrapWithReduxAndStyle from "../../testHelpers/wrapWithReduxAndStyle";
import GridColumn from "../GridColumn";

const gridColumnText = "column";

describe("<GridColumn />", (): void => {
  it("renders <GridColumn /> component", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<GridColumn>{gridColumnText}</GridColumn>));
    expect(container).toMatchSnapshot();
  });

  it("renders <GridColumn /> in bleed variant", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<GridColumn bleed>{gridColumnText}</GridColumn>));
    const element = getByText(gridColumnText);
    expect(element).toHaveStyleRule("padding", "0");
  });

  it("renders <GridColumn /> in half size", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<GridColumn size="half">{gridColumnText}</GridColumn>));
    const element = getByText(gridColumnText);
    expect(element).toHaveStyleRule("width", "calc(100% / (12/6))");
  });

  it("renders <GridColumn /> in size 3", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<GridColumn size={3}>{gridColumnText}</GridColumn>));
    const element = getByText(gridColumnText);
    expect(element).toHaveStyleRule("width", "calc(100% / (12/3))");
  });

  it("renders <GridColumn /> in size 3 with offset 3", (): void => {
    const { getByText } = render(
      wrapWithReduxAndStyle(
        <GridColumn size={3} offset={3}>
          {gridColumnText}
        </GridColumn>
      )
    );
    const element = getByText(gridColumnText);
    expect(element).toHaveStyleRule("margin-left", "calc(100% / (12/3))");
  });

  it("renders <GridColumn /> in size 4 and medium 3 and large 2 and huge 1", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <GridColumn size={4} sizeMedium={3} sizeLarge={2} sizeHuge={1}>
          {gridColumnText}
        </GridColumn>
      )
    );
    expect(container).toMatchSnapshot();
  });

  it("renders <GridColumn /> in size 4, offsets: 4 and medium 3 and large 2 and huge 1", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <GridColumn size={4} offset={4} offsetMedium={3} offsetLarge={2} offsetHuge={1}>
          {gridColumnText}
        </GridColumn>
      )
    );
    expect(container).toMatchSnapshot();
  });
});
