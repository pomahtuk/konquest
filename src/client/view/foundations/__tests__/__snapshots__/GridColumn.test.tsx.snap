// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridColumn /> renders <GridColumn /> component 1`] = `
.emotion-0 {
  box-sizing: border-box;
  padding: 8px calc(8px/2) 0 calc(8px/2);
  -webkit-flex-basis: calc(100% / (12/12));
  -ms-flex-preferred-size: calc(100% / (12/12));
  flex-basis: calc(100% / (12/12));
  width: calc(100% / (12/12));
  max-width: calc(100% / (12/12));
}

@media screen and (min-width:576px) {
  .emotion-0 {
    padding: 16px calc(16px/2) 0 calc(16px/2);
  }
}

@media screen and (min-width:992px) {
  .emotion-0 {
    padding: 24px calc(24px/2) 0 calc(24px/2);
  }
}

@media screen and (min-width:1200px) {
  .emotion-0 {
    padding: 32px calc(32px/2) 0 calc(32px/2);
  }
}

<div>
  <div
    class="emotion-0"
  >
    column
  </div>
</div>
`;

exports[`<GridColumn /> renders <GridColumn /> in size 4 and medium 3 and large 2 and huge 1 1`] = `
.emotion-0 {
  box-sizing: border-box;
  padding: 8px calc(8px/2) 0 calc(8px/2);
  -webkit-flex-basis: calc(100% / (12/4));
  -ms-flex-preferred-size: calc(100% / (12/4));
  flex-basis: calc(100% / (12/4));
  width: calc(100% / (12/4));
  max-width: calc(100% / (12/4));
}

@media screen and (min-width:576px) {
  .emotion-0 {
    padding: 16px calc(16px/2) 0 calc(16px/2);
    width: calc(100% / (12/3));
    -webkit-flex-basis: calc(100% / (12/3));
    -ms-flex-preferred-size: calc(100% / (12/3));
    flex-basis: calc(100% / (12/3));
    max-width: calc(100% / (12/3));
  }
}

@media screen and (min-width:992px) {
  .emotion-0 {
    padding: 24px calc(24px/2) 0 calc(24px/2);
    width: calc(100% / (12/2));
    -webkit-flex-basis: calc(100% / (12/2));
    -ms-flex-preferred-size: calc(100% / (12/2));
    flex-basis: calc(100% / (12/2));
    max-width: calc(100% / (12/2));
  }
}

@media screen and (min-width:1200px) {
  .emotion-0 {
    padding: 32px calc(32px/2) 0 calc(32px/2);
    width: calc(100% / (12/1));
    -webkit-flex-basis: calc(100% / (12/1));
    -ms-flex-preferred-size: calc(100% / (12/1));
    flex-basis: calc(100% / (12/1));
    max-width: calc(100% / (12/1));
  }
}

<div>
  <div
    class="emotion-0"
  >
    column
  </div>
</div>
`;

exports[`<GridColumn /> renders <GridColumn /> in size 4, offsets: 4 and medium 3 and large 2 and huge 1 1`] = `
.emotion-0 {
  box-sizing: border-box;
  padding: 8px calc(8px/2) 0 calc(8px/2);
  -webkit-flex-basis: calc(100% / (12/4));
  -ms-flex-preferred-size: calc(100% / (12/4));
  flex-basis: calc(100% / (12/4));
  width: calc(100% / (12/4));
  max-width: calc(100% / (12/4));
  margin-left: calc(100% / (12/4));
}

@media screen and (min-width:576px) {
  .emotion-0 {
    padding: 16px calc(16px/2) 0 calc(16px/2);
    margin-left: calc(100% / (12/3));
  }
}

@media screen and (min-width:992px) {
  .emotion-0 {
    padding: 24px calc(24px/2) 0 calc(24px/2);
    margin-left: calc(100% / (12/2));
  }
}

@media screen and (min-width:1200px) {
  .emotion-0 {
    padding: 32px calc(32px/2) 0 calc(32px/2);
    margin-left: calc(100% / (12/1));
  }
}

<div>
  <div
    class="emotion-0"
  >
    column
  </div>
</div>
`;
